### DISCUSSION TITLE
PipesClient implementation

### DISCUSSION THREAD

**Discussion by krgn** [dagster-io/dagster#18215] [Q&A] [answered]
Hi all,

I have hit an issue that I believe could be of interest to a wider audience, hence I'm opening a discussion here. 

While implementing some pipes jobs, the need arose to be able to switch out how things are run depending on the environment, i.e. running the job in a `PipesSubprocessClient` vs. `PipesK8sClient` for local testing. This was more challenging that I believe it needs to be, or should be, for that matter.

I solved it like this now:

```python
class MyPipesClient(ConfigurableResource):
    @abstractmethod
    def run(
        self,
         ....  # all the params needed to instantiate a job
     ):
        pass


class _K8sClient(MyPipesClient):
    def __init__(self, client: PipesK8sClient):
        self.client = client

    def run(
        self,
         ....  # all the params needed to instantiate a job
     ):
        # construct all necessary dependencies for PipesK8sClient
        self.client.run(...) 

class _LocalClient(MyPipesClient):
    def run(
        self,
         ....  # all the params needed to instantiate a job
     ):
        client = PipesSubprocessClient()
        client.run(...) 


K8sClient = ResourceParam[_K8sClient]
LocalClient = ResourceParam[_LocalClient]
```

Finally, using `MyPipesClient` as the type in function signatures works. 

It took me several attempts to figure this out, so I would like to pose the following questions:

* is this the recommended approach to generalize over the client type?
* would it make sense to make `PipesClient` public, so that it could be used in function signatures and ease this process somewhat?
* should this workflow be documented somewhere? (I've looked at several pages regarding resources, never quite finding what I was looking for)

Thank you for reading this far! 🙂

**Comment by schrockn** [comment, marked as answer]
Hey @krgn I agree that this is too difficult right now.

For some context, you do _not_ need to inherit from `ConfigurableResource` in order to use the resource system for dependency injection. This can be done with raw Python objects. Configuration is needed if you want to:

1. Change the configuration of execution at runtime (via the launchpad)
2. View/observe configuration in the UI
3. Use environment variables to store secrets.

I think your request to make `PipesClient` public is very reasonable to handle this use case. Another option to explore the use of [Protocols](https://peps.python.org/pep-0544/).

We shouldn't force you `_LocalClient` pattern that delegates to an inner client, as that is quite clunky.

cc: @alangenfeld

**Reply by krgn** [reply]
Thank you for the explanation @schrockn, that clears things up for me (esp. the usage/intention of `ConfigurableResource`). I just did a few more tests, and the simplest solution is to do it seems to be this:

```python
class _MyPipesClient:
    @abstractmethod
    def run(self, ...):
        """Abstract!"""

MyPipesClient = ResourceParam[_MyPipesClient]

class K8sClient(MyPipesClient):
    def __init__(self, client: PipesK8sClient):
        self.client = client

    def run(self, ...):
        # construct all necessary dependencies for PipesK8sClient
        self.client.run(...) 

class LocalClient(MyPipesClient):
    def run(self, ...):
        client = PipesSubprocessClient()
        client.run(...) 
```

I will explore Protocols, too, they look like a good pattern for this use-case. 

Thanks!

**Comment by alangenfeld** [comment]
> switch out how things are run depending on the environment, i.e. running the job in a PipesSubprocessClient vs. PipesK8sClient for local testing

So for this specific case, what exactly do you invoke in a subprocess given the arguments that are typically used to run a kubernetes pod? Do you have some assumptions that allow you to translate to running an equivalent computation locally instead of via a container or do you just run some dummy subprocess?

**Reply by krgn** [reply]
In this particular case, all the machinery that pertains to the runtime environment are hidden inside the protocol implementations (e.g. how to construct the pods, volumes, env vars etc in the case of k8s, or, alternatively, how to run the workload in a sub-process). In practice, both environments share the same runner script, and just configure it differently. Does that answer your question?

**Reply by alangenfeld** [reply]
Yep thanks for the details.
